Adding eval to the SFT

HosseinKaviani-H opened the pull request:
# Add periodic evaluation during training with epoch-aware synchronization

Added evaluation functionality to the SFT training recipe with proper multi-rank synchronization and epoch completion detection.

## Changes:

### Core Evaluation Features

* **Configurable evaluation interval**: Added `eval_interval` and `eval_steps` parameters to control when and how much to evaluate
  * `eval_interval`: Number of training steps between evaluations (defaults to `float('inf')` to disable eval when not configured)
  * `eval_steps`: Number of validation batches to evaluate per evaluation run (defaults to `0` for unlimited - runs one full epoch)

* **Validation dataloader**: Set up separate validation dataloader using the last 10% of the train split

* **Forward-only pass**: Implemented `forward_only()` method for evaluation without gradient computation, supporting both pipeline parallel and non-PP configurations

### Epoch-Aware Evaluation with Multi-Rank Synchronization

* **Epoch completion detection**: Evaluates for exactly one complete epoch by monitoring `batch["metrics"]` for epoch increments
  * Extracts `num_epochs` from batch metadata to detect when validation dataset completes one full pass
  * Prevents evaluation from running forever on infinite streaming dataloaders

* **Non-blocking all_reduce pattern**: Synchronizes epoch completion across all ranks without blocking computation
  * **Prefetch + async pattern**: Fetches next batch while processing current batch
  * **Overlapped communication**: Starts `async_op=True` all_reduce on next batch's epoch while GPU computes current batch's loss
  * **Deferred checking**: Checks all_reduce result in next iteration (should be complete by then)
  * **Early stopping**: All ranks stop when ANY rank completes an epoch (via MAX reduction)

### Integration

* **Training integration**: Evaluation runs automatically every `eval_interval` steps during training
* **Graceful degradation**: If `eval_steps > 0`, it acts as a cap (useful for quick validation checks or when epoch metadata is unavailable)

## Usage:

Configure in your YAML config file:

```yaml
training:
  eval_interval: 100  # Evaluate every 100 training steps
  eval_steps: 0       # Run one complete epoch (recommended)
  # eval_steps: 50    # Alternative: cap at 50 batches for faster checks
```
If [eval_interval](command:code-compose.open?%5B%22%2Fhome%2Fhosseinkh%2Fforge_updated%2Fforge%2Fapps%2Fsft%2Fmain.py%22%2C83%5D)and [eval_steps](command:code-compose.open?%5B%22%2Fhome%2Fhosseinkh%2Fforge_updated%2Fforge%2Fapps%2Fsft%2Fmain.py%22%2C84%5D) are not set, evaluation is automatically disabled.

Testing:
Comprehensive test suite (test_evaluate.py) validates:

✅ Epoch extraction from batch metadata
✅ Single epoch completion detection
✅ eval_steps cap enforcement
✅ Empty/single batch edge cases
✅ Async all_reduce pattern behavior
✅ Multi-rank synchronization logic
✅ Prefetch pattern correctness
All 14 tests pass successfully.

Algorithm Details:
The non-blocking evaluation loop follows this pattern:

Iteration N:
  1. Check if previous all_reduce says we should stop
  2. Process current batch (forward pass, compute loss)
  3. Prefetch next batch
  4. Extract epoch from next batch
  5. Start async all_reduce on epoch_increment (returns immediately, doesn't block)
  
Iteration N+1:
  1. Wait for all_reduce from iteration N (should be done, or very fast)
  2. Check result: if any rank has epoch_increment > 0, stop
  3. Process batch N+1...
This overlaps network communication with GPU computation for better performance, while ensuring all ranks stop at the same point.


This updated description captures:
1. The epoch detection mechanism
2. The async all_reduce implementation details
3. The performance benefits of the overlapped communication pattern
4. Test coverage
5. Clear usage examples
6. Algorithm explanation for reviewers

HamidShojanazeri commented on apps/sft/llama3_8b.yaml:
```diff
@@ -33,6 +33,8 @@ training:
   steps: 1000
   compile: false
   dataset: "c4"
```
would the dataset be `c4` or alpaca?

HosseinKaviani-H commented:
> hey @HosseinKaviani-H , thanks for opening the PR. Its a bit tricky to run the validation, because the dataset is infinite. So it doesnt know when to stop. You can retrieve the epoch number for each dataset from batch["metrics'], but we haven't looked into that. On top of that, if you have multiple datasets, they will epoch at different paces. I think that there are a few ways on handling this:
> 
> 1. changing the dataset definition for validation / adding some flag / some logic that goes in the dataset
> 2. Trying to orchestrate it in the recipe, i.e. checking if the dataset has epoched and stopping then.
> 
> It seems that you defined "eval_steps" as a clever solution to not deal with any of that. But i wonder about correctness here, i.e. not going through the entire eval, or going 1.5x times, for example. Any thoughts?

Hi Felipe,

Thanks for your comment. Yeah I think one good solution as you mentioned is to retrieve the epoch number in the training loop and once it hits 0 to 1, it breaks. I'll try to give it some thoughts and implement it. 

And yes, counting batches is arbitrary here as if eval_steps is too low it could lead to incomplete evaluation or too high it might cause duplicate evaluation. Hence, checking epoch number sounds a better solution here.

ebsmothers commented:
@felipemello1 sorry maybe I don't fully understand your suggestions then. What is the termination condition for the validation loop? If it is epoch-based in any way I think we will run into this issue, right?

ebsmothers commented on apps/sft/main.py:
```diff
@@ -111,25 +113,23 @@ def _init_dist(self):
-        # Should also use load, not dcp_load
+        # Setup training data (first 90% of train split)
+        self.train_dataloader = self.setup_data(
+            dataset_path="yahma/alpaca-cleaned", dataset_split="train[:90%]"
```
I didn't realize we were hardcoding the dataset in this one. Can we adopt the changes from #54 here to make it configurable (I think this will also address Hamid's comment above)?

ebsmothers commented on apps/sft/main.py:
```diff
@@ -146,8 +146,8 @@ def setup_data(self):
         dataset = sft_iterable_dataset(
             model_transform=tokenizer,
             message_transform=AlpacaToMessages(),
```
Ideally we shouldn't hardcode this either (but it's a bit more work without instantiate)

HosseinKaviani-H replied:
I agree. We can have this implemented as soon as we fix the main eval functionality

ebsmothers commented on apps/sft/main.py:
```diff
@@ -254,18 +347,21 @@ async def train(self) -> None:
 
+            # Run evaluation periodically
+            if self.current_step % self.eval_interval == 0:
+                eval_metrics = await self.evaluate()
```
Noob q: what is the implication of using await here? Doesn't this return control to the parent process (which in this case would be `run`)? Does this actually buy us anything?

HosseinKaviani-H commented:
> # Add periodic evaluation during training with epoch-aware synchronization
> Added evaluation functionality to the SFT training recipe with proper multi-rank synchronization and epoch completion detection.
> 
> ## Changes:
> ### Core Evaluation Features
> * **Configurable evaluation interval**: Added `eval_interval` and `eval_steps` parameters to control when and how much to evaluate
>   
>   * `eval_interval`: Number of training steps between evaluations (defaults to `float('inf')` to disable eval when not configured)
>   * `eval_steps`: Number of validation batches to evaluate per evaluation run (defaults to `0` for unlimited - runs one full epoch)
> * **Validation dataloader**: Set up separate validation dataloader using the last 10% of the train split
> * **Forward-only pass**: Implemented `forward_only()` method for evaluation without gradient computation, supporting both pipeline parallel and non-PP configurations
> 
> ### Epoch-Aware Evaluation with Multi-Rank Synchronization
> * **Epoch completion detection**: Evaluates for exactly one complete epoch by monitoring `batch["metrics"]` for epoch increments
>   
>   * Extracts `num_epochs` from batch metadata to detect when validation dataset completes one full pass
>   * Prevents evaluation from running forever on infinite streaming dataloaders
> * **Non-blocking all_reduce pattern**: Synchronizes epoch completion across all ranks without blocking computation
>   
>   * **Prefetch + async pattern**: Fetches next batch while processing current batch
>   * **Overlapped communication**: Starts `async_op=True` all_reduce on next batch's epoch while GPU computes current batch's loss
>   * **Deferred checking**: Checks all_reduce result in next iteration (should be complete by then)
>   * **Early stopping**: All ranks stop when ANY rank completes an epoch (via MAX reduction)
> 
> ### Integration
> * **Training integration**: Evaluation runs automatically every `eval_interval` steps during training
> * **Graceful degradation**: If `eval_steps > 0`, it acts as a cap (useful for quick validation checks or when epoch metadata is unavailable)
> 
> ## Usage:
> Configure in your YAML config file:
> 
> ```yaml
> training:
>   eval_interval: 100  # Evaluate every 100 training steps
>   eval_steps: 0       # Run one complete epoch (recommended)
>   # eval_steps: 50    # Alternative: cap at 50 batches for faster checks
> ```
> 
> If eval_intervaland eval_steps are not set, evaluation is automatically disabled.
> 
> Testing: Comprehensive test suite (test_evaluate.py) validates:
> 
> ✅ Epoch extraction from batch metadata ✅ Single epoch completion detection ✅ eval_steps cap enforcement ✅ Empty/single batch edge cases ✅ Async all_reduce pattern behavior ✅ Multi-rank synchronization logic ✅ Prefetch pattern correctness All 14 tests pass successfully.
> 
> Algorithm Details: The non-blocking evaluation loop follows this pattern:
> 
> Iteration N:
> 
> 1. Check if previous all_reduce says we should stop
> 2. Process current batch (forward pass, compute loss)
> 3. Prefetch next batch
> 4. Extract epoch from next batch
> 5. Start async all_reduce on epoch_increment (returns immediately, doesn't block)
> 
> Iteration N+1:
> 
> 1. Wait for all_reduce from iteration N (should be done, or very fast)
> 2. Check result: if any rank has epoch_increment > 0, stop
> 3. Process batch N+1...
>    This overlaps network communication with GPU computation for better performance, while ensuring all ranks stop at the same point.
> 
> This updated description captures:
> 
> 1. The epoch detection mechanism
> 2. The async all_reduce implementation details
> 3. The performance benefits of the overlapped communication pattern
> 4. Test coverage
> 5. Clear usage examples
> 6. Algorithm explanation for reviewers

@ebsmothers @felipemello1 Given our discussion and per Felipe's idea, I have implemented an epoch-based eval with non-blocking all-reduce. I have updated the description and added a test_evaluate script to cover different scenarios.

HosseinKaviani-H commented on apps/sft/main.py:
```diff
@@ -254,18 +347,21 @@ async def train(self) -> None:
             for k, v in batch.items():
                 if isinstance(v, torch.Tensor):
-                    batch[k] = v.to("cuda")  # TODO: hardcoded for now
+                    batch[k] = v.to(self.device)  # TODO: hardcoded for now
```
Fixed

felipemello1 commented on apps/sft/main.py:
```diff
@@ -225,27 +243,173 @@ def forward_backward(
+                    next_batch = next(val_dataloader)
+                    next_epoch = self._extract_epoch_from_batch(next_batch)
+
+                    if next_epoch is not None and starting_epoch is not None:
```
i am tempted to say lets remove the if/else check for None. Do you see a strong argument for keeping them?

HosseinKaviani-H replied:
That's true for the datasets that have epoch metrics, but if user adds arbitrary datasets, we need this safeguard.

felipemello1 commented on apps/sft/main.py:
```diff
@@ -225,27 +243,173 @@ def forward_backward(
+        for model_part in self.model_parts:
+            model_part.train()
+
+        avg_loss = total_loss / max(num_batches, 1)
```
This might be incorrect. If we have one sample with 10 tokens and another sample with 100 tokens, we have to /110, not by 2. @ebsmothers  can you confirm?

felipemello1 commented on apps/sft/main.py:
```diff
@@ -225,27 +243,173 @@ def forward_backward(
+        for model_part in self.model_parts:
+            model_part.eval()
+
+        val_dataloader = iter(self.val_dataloader)
```
did you have a chance to consider the case for multidataset? what happens in this case?

felipemello1 reviewed (commented):
Sorry for the delay. Left some comments/suggestions. We would need to test it in some distributed capacity. Were you able to run it for >1 node and confirm that it stopped right after 1 epoch?

ebsmothers commented on apps/sft/main.py:
```diff
@@ -111,25 +129,30 @@ def _init_dist(self):
+        dataset_val_config = self.job_config.get("dataset_val", {})
+        self.val_dataloader = self.setup_data(
+            dataset_path=dataset_val_config.get("path", dataset_config.get("path")),
+            dataset_split=dataset_val_config.get("split", dataset_config.get("split")),
```
nit but I don't like these nested .get calls. It also seems strange that we would fallback to validation on the training set. Personally I would just recommend checking if validation is enabled, and if it's not, don't even set up the validation dataloader at all.

HosseinKaviani-H commented on apps/sft/main.py:
```diff
@@ -225,27 +243,173 @@ def forward_backward(
+        # Prefetch first batch
+        try:
+            next_batch = next(val_dataloader)
+        except StopIteration:
```
I have removed the defensive checks here.
